src/components/AutoComplete.jsx

```diff
@@ -0,0 +1,584 @@
+    }
+  }
+
+  /**
+   * The function is called when user inputs a value in the input box. Function will do following:
+   * 1. Open the options
+   * 2. Change value of input depending on whether its has value, defaultValue or valueLink property
+   * 3. Call onChange props function
+   */
+  _onChange(event) {
+    const value = event.target.value;
+    if(has(this.props, 'valueLink')) {
+      this.props.valueLink.requestChange(value);
+      this.setState({
+        isOpen: true,
+        focusedOptionIndex: 0
+      });
+    }
+    else if(has(this.props, 'value')) {
+      this.setState({
+        isOpen: true,
+        focusedOptionIndex: 0
+      });
+    }
+    else {
+      this.setState({
+        inputValue: value,
+        isOpen: true,
+        focusedOptionIndex: 0,
+        filteredOptions: this.filterOptions(value)
+      });
+    }
+
+    const domNode = React.findDOMNode(this);
+    domNode.value = value;
+
+    if (this.props.onChange) {
+      this.props.onChange({target: domNode});
+    }
+  }
+
+  /**
+   * Function to filter options using input value.
+   */
+  filterOptions(inputValue) {
+    let filteredOptions = [];
+    if(this.props.children.length > 0) {
+      if(inputValue) {
+        filteredOptions = filter(this.props.children, (entry) => {
+          return this.props.filterFunc(inputValue, entry.props.value);
+        });
+      } else {
+        filteredOptions = map(this.props.children, (entry) => { return entry;});
+      }
+      if(this.props.maxOptions) {
+        filteredOptions = filteredOptions.splice(0, this.props.maxOptions);
+      }
+    }
+    return filteredOptions;
+  }
+
+  render () {
+
+    const wrapperStyle = extend({}, style.wrapperStyle, this.props.wrapperStyle);
+    let inputStyle = extend({}, style.style, this.props.style);
+    const menuStyle = extend({}, style.menuStyle, this.props.menuStyle);
+
+    const inputClassName = unionClassNames(this.props.className, this._styleId);
+
+    const tabIndex = this.props.tabIndex ? this.props.tabIndex : '0';
+
+    // Currently there are no different hover styles for caret, like select they are probably not really needed.
+    if(this.props.displayCaret) {
+      if (this.props.disabled) {
+        inputStyle = extend({}, inputStyle, style.disabledCaretToOpenStyle);
+      } else if (this.state.isOpen) {
+        inputStyle = extend({}, inputStyle, style.caretToCloseStyle);
+      } else {
+        inputStyle = extend({}, inputStyle, style.caretToOpenStyle);
+      }
+    }
+
+    const computedMenuStyle = (this.state.isOpen && !this.props.disabled && this.state.filteredOptions && this.state.filteredOptions.length > 0) ? menuStyle : { display: 'none' };
+
+    //using value for input makes it a controlled component and it will be changed in controlled manner if (1) user enters value, (2) user selects some option
+    //value will be updated depending on whether user has passed value / valueLink / defaultValue as property
```

Comment:
Like select I have added ref to all <div>, <input> and <ul>, I am not sure but user may need these.